# DS-SF-36 | 03 | `pandas` | Assignment | Starter Code

## `pandas` and Bistro

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `pandas` (as `pd`).

In [3]:
import os

import numpy as np
import pandas as pd
# TODO

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

> ### Question 2.  Read the `dataset-03-bistro.csv` dataset.

In [7]:
# TODO
!ls ../datasets/dataset-03-bistro.csv
# !cat ../datasets/dataset-03-bistro.csv

../datasets/dataset-03-bistro.csv


> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [10]:
# TODO
df = pd.read_csv(os.path.join('..', 'datasets', 'dataset-03-bistro.csv'))
type(df)

pandas.core.frame.DataFrame

Answer: dataframe

> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [11]:
# TODO
df.shape[0]

244

Answer: 244

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [12]:
# TODO
df.shape[1]

8

Answer: 8

> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [13]:
# TODO
df.dtypes

day           object
time          object
name          object
gender        object
is_smoker       bool
party          int64
check        float64
tip          float64
dtype: object

> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [19]:
# TODO
df.head(2)
# df.loc[:1]

day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66

Answer: TODO

> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [20]:
# TODO
df.tail(2)

day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

Answer: TODO

> ### Question 9.  Does the dataset contain any missing values?

In [21]:
# TODO
df.isnull().sum()


day          0
time         0
name         0
gender       0
is_smoker    0
party        0
check        0
tip          0
dtype: int64

Answer: No empty values

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [36]:
# TODO
smoker = df[['is_smoker']]
df.groupby(['is_smoker']).size()
# Since no one smokes we can drop this column
df.drop(['is_smoker'], axis = 1, inplace = True)
df.head()


day    time      name  gender  party  check   tip
0  Sunday  Dinner  Kimberly  Female      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      3  10.34  1.66
2  Sunday  Dinner     Larry    Male      3  21.01  3.50
3  Sunday  Dinner    Joseph    Male      2  23.68  3.31
4  Sunday  Dinner    Janice  Female      4  24.59  3.61

Answer: Since there are no smokers in the dataset, we dropped the is_smoker column

> ### Question 11.  For which week days does the dataset has data for?

In [39]:
# TODO
df.groupby(['day']).size()

day
Friday      19
Saturday    87
Sunday      76
Thursday    62
dtype: int64

Answer: Thursday, Friday, Saturday, Sunday

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [40]:
# TODO
# df[['col1', 'col2', 'col3', 'col4']].groupby(['col1', 'col2']).agg(['mean', 'count'])
df.groupby(['day']).size()


day
Friday      19
Saturday    87
Sunday      76
Thursday    62
dtype: int64

Answer: TODO

> ### Question 13.  How much tip did waiters collect for each week day?

In [43]:
# TODO
df[['day', 'tip']].groupby(['day']).agg(['sum'])

tip
             sum
day             
Friday     51.96
Saturday  260.40
Sunday    247.39
Thursday  171.83

Answer: TODO

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [44]:
# TODO
df[['day', 'tip']].groupby(['day']).agg(['mean'])

tip
              mean
day               
Friday    2.734737
Saturday  2.993103
Sunday    3.255132
Thursday  2.771452

Answer: TODO

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [56]:
# TODO

df['percentCheck'] = df['tip']/df['check']
df
df[['day', 'check', 'tip', 'percentCheck']].groupby(['day']).agg(['mean'])


check       tip percentCheck
               mean      mean         mean
day                                       
Friday    17.151579  2.734737     0.169913
Saturday  20.441379  2.993103     0.153152
Sunday    21.410000  3.255132     0.166897
Thursday  17.682742  2.771452     0.161276

Answer: TODO

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [95]:
# TODO
# List of male names
males = df[df.gender == 'Male']

# List of female names
females = df[df.gender == 'Female']

# List of intersections
commonNames = np.intersect1d(males['name'], females['name'])
commonNames

array(['Casey'], dtype=object)

Answer: Casey

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [96]:
# TODO
names = df['name'].unique()
print names
length = len(names)
print length

# We could add the number of common names among men and women ('Casey) to be more correct


['Kimberly' 'Nicholas' 'Larry' 'Joseph' 'Janice' 'Mark' 'Jamie' 'Steven'
 'Wilburn' 'Ian' 'Rick' 'Katie' 'Karl' 'Timothy' 'Anne' 'Brandon' 'Alice'
 'Roman' 'Paisley' 'Sean' 'William' 'Pauline' 'Jo' 'Richard' 'Carl' 'Mario'
 'Charles' 'Tyler' 'Jonathan' 'Jan' 'Dustin' 'Reyes' 'Jessica' 'Laura'
 'Alejandro' 'Levi' 'Leonard' 'Lori' 'Casey' 'Gregory' 'James' 'Arturo'
 'Maurice' 'Melvin' 'David' 'Chandler' 'Frank' 'Harold' 'Keith' 'Jennie'
 'Wanda' 'Brian' 'Christopher' 'Marvin' 'Destiny' 'Jerry' 'Diego' 'Bryson'
 'Matt' 'Daniel' 'Harry' 'Rebecka' 'Bailey' 'Ronald' 'Jocelyn' 'Victoria'
 'Lisa' 'Miriam' 'Andrew' 'Edwin' 'Erick' 'Anna' 'Kaleb' 'Donna' 'Brent'
 'Kenneth' 'Todd' 'Maverick' 'Brandi' 'Melinda' 'Eleanor' 'Curtis' 'Kevin'
 'Alex' 'Roger' 'Adalberto' 'Martha' 'Dora' 'Mildred' 'Tracy' 'Dena' 'Kody'
 'Kelly' 'Theresa' 'Javier' 'Ruth' 'Michal' 'Erin' 'Raelynn' 'Imani'
 'Stanley' 'Lena' 'Shaquasia' 'Mary' 'Norman' 'Karen' 'Walter' 'Elizabeth'
 'Marian' 'Enrique' 'Lewis' 'Jaycee' 'Brenda

Answer: 181

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [97]:
# TODO
kevin = df[df.name == 'Kevin']
print "Number of times Kevin patronized: ", len(kevin)
alice = df[df.name == 'Alice']
print "Number of times Alice patronized: ", len(alice)

Number of times Kevin patronized:  4
Number of times Alice patronized:  2


Answer: TODO

> ### Question 19.  Who are the top 3 female and male patrons?

In [113]:
# TODO
maleFreq = males.groupby(['name']).size()
maleFreq.sort_values(axis=0, ascending=False, inplace=True, kind='quicksort', na_position='last')
topMales = maleFreq.head(3)
print topMales

femaleFreq = females.groupby(['name']).size()
femaleFreq.sort_values(axis=0, ascending=False, inplace=True, kind='quicksort', na_position='last')
topFemales = femaleFreq.head(3)
print topFemales
# df.groupby(['day']).size()

name
David    8
James    5
Casey    5
dtype: int64
name
Mary     4
Casey    3
Laura    3
dtype: int64


Answer: David, James, Casey ; Mary, Casey, Laura


> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [126]:
# TODO
best = df.sort_values(by='percentCheck', axis=0, ascending=False, inplace=False, kind='quicksort', na_position='last')
print best.head(1)
# worst = df.sort_values(by='percentCheck', axis=0, ascending=True, inplace=False, kind='quicksort', na_position='last')
print best.tail(1)

daniel = df[df.name == 'Daniel']
print "Number of times Daniel patronized: ", len(daniel)
jeremy = df[df.name == 'Jeremy']
print "Number of times Jeremy patronized: ", len(jeremy)

         day    time    name gender  party  check   tip  percentCheck
172  Sunday  Dinner  Daniel   Male      2   7.25  5.15      0.710345
          day    time    name gender  party  check   tip  percentCheck
237  Saturday  Dinner  Jeremy   Male      2  32.83  1.17      0.035638
Number of times Daniel patronized:  2
Number of times Jeremy patronized:  1


Answer: TODO